drivers/hwmon_ina219.c

```diff
@@ -0,0 +1,496 @@
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with this program; if not, write to the Free Software
+	Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+ */
+
+#include "main.h"
+#include "nut_float.h"
+#include "nut_stdint.h"
+
+#include <dirent.h>
+#include <errno.h>
+#include <string.h>
+#include <sys/types.h>
+
+#ifndef STRFY0
+	#define STRFY0(x) #x
+#endif
+#ifndef STRFY
+	#define STRFY(x) STRFY0(x)
+#endif
+
+#define SYSFS_HWMON_DIR                     "/sys/class/hwmon"
+#define BATTERY_CHARGE_LOW                  15
+#define BATTERY_CHARGE_HIGH                 75
+#define DRIVER_NAME                         "hwmon-INA219 UPS driver"
+#define DRIVER_VERSION                      "0.01"
+
+upsdrv_info_t upsdrv_info = {
+	DRIVER_NAME,
+	DRIVER_VERSION,
+	"Jan Viktorin <jan.viktorin@gmail.com>",
+	DRV_EXPERIMENTAL,
+	{ NULL },
+};
+
+/**
+ * @brief Path usually pointing to /sys/class/hwmon/hwmonX.
+ */
+static char ina219_base_path[PATH_MAX];
+
+/**
+ * @brief Threshold for detection of LB status.
+ */
+static unsigned int battery_charge_low = BATTERY_CHARGE_LOW;
+
+/**
+ * @brief Threshold for detection of HB status.
+ */
+static unsigned int battery_charge_high = BATTERY_CHARGE_HIGH;
+
+/**
+ * @brief Battery voltage (mV) when it is considered depleted.
+ */
+static unsigned int battery_voltage_min;
+
+/**
+ * @brief Battery voltage (mV) when it is considered fully charged.
+ */
+static unsigned int battery_voltage_max;
+
+/**
+ * @brief Voltage value as recently read from in1_input (mV).
+ * @see https://docs.kernel.org/hwmon/ina2xx.html
+ */
+static int voltage = 0;
+
+/**
+ * @brief Current value as recently read from curr1_input (mA).
+ * @see https://docs.kernel.org/hwmon/ina2xx.html
+ */
+static int current = 0;
+
+static int file_contains(const char *path, const char *text)
+{
+	FILE *f;
+	char buf[128];
+	size_t len;
+
+	if ((f = fopen(path, "r")) == NULL) {
+		upslog_with_errno(LOG_ERR, "unable to open %s", path);
+		return 0;
+	}
+
+	len = fread(buf, 1, sizeof(buf), f);
+	fclose(f);
+	upsdebugx(4, "read %" PRIuSIZE " bytes", len);
+
+	if (strlen(text) != len)
+		return 0;
+
+	if (memcmp(buf, text, len))
+		return 0;
+
+	return 1;
+}
+
+static int file_read_number(const char *path, int *value)
+{
+	char buf[128];
+	ssize_t ret;
+	int fd;
+
+	if ((fd = open(path, O_RDONLY)) < 0)
+		return -errno;
+
+	if ((ret = pread(fd, buf, sizeof(buf), 0)) < 0) {
+		const int _e = -errno;
+		close(fd);
+		return _e;
+	}
+	else {
+		close(fd);
+
+		buf[ret] = '\0';
+		*value = atoi(buf);
+		return 0;
+	}
+}
+
+static int check_ina219(const char *ina219_dir)
+{
+	char namepath[PATH_MAX];
+
+	upsdebugx(3, "checking %s", ina219_dir);
+
+	snprintf(namepath, sizeof(namepath), "%s/name", ina219_dir);
+
+	if (!file_contains(namepath, "ina219\n"))
+		return -ENODEV;
+
+	upsdebugx(3, "detected ina219 at %s", ina219_dir);
+	return 0;
```

Comment:
Hm, this (and the use below) is a bit un-idiomatic, especially for a method named like a verb (`check` something) to return a not-true value when the check is successful.  I guess it is due to a non-zero negative error as a possible return.

Not sure if there's a good way out, maybe negate the name if it's hard to negate the return value (so below you would go `if (check_not_ina219()) continue;` for non-INA cases)...

Reply:
Ok, the goal was not to return bool but to return OK or error code. I consider names `check_*()` as non-bool and expect them to return something more informative. For bool calls, I am used to name functions as `is_*()` or `can_*()`. But, no problem to rename it.

I would suggest to rename `detect_hwmon_ina219()` to `scan_hwmon_ina219()` and this `check_ina219()` function to `detect_hwmon_ina219()`. Does it make sense?

Reply:
I think so, thanks.

Reply:
Renamed.